mapedit/studio.cc

```diff
@@ -3190,8 +3208,45 @@ void convertFromUTF8::convert(gchar *&_convstr, const char *str, const char *enc
 	_convstr = g_convert(str, -1, enc, "UTF-8",
 	                     &bytes_read, &bytes_written, &error);
 
-	if (_convstr)
+	if (_convstr) {
+		if (is_7bit) {
+			for (unsigned char *ch = reinterpret_cast<unsigned char *>(_convstr); *ch; ch++) {
+				if      (*ch <= 0x7f) ;
+				// 80 -> 01 Ç:C cedilla,    81 -> 02 ü:u diaeresis,  82 -> 03 é:e acute,
+				// 83 -> 04 â:a circumflex, 84 -> 05 ä:a diaeresis,  85 -> 06 à:a grave.
+				else if (*ch <= 0x85) *ch -= 0x7f;
+				else if (*ch <= 0x86) ;
+				// 87 -> 07 ç:c cedilla,    88 -> 08 ê:e circumflex.
+				else if (*ch <= 0x88) *ch -= 0x80;
+				// 89 -> 0b ë:e diaeresis,  8a -> 0c è:e grave.
+				else if (*ch <= 0x8a) *ch -= 0x7e;
+				// 8b -> 0e ï:i diaeresis,  8c -> 0f î:i circumflex, 8d -> 10 ì:i grave,
+				// 8e -> 11 Ä:A diaeresis.
+				else if (*ch <= 0x8e) *ch -= 0x7d;
+				else if (*ch <= 0x8f) ;
+				// 90 -> 12 É:E acute E.
+				else if (*ch <= 0x90) *ch -= 0x7e;
+				else if (*ch <= 0x92) ;
+				// 93 -> 13 ô:o circumflex, 94 -> 14 ö:o diaeresis.
+				else if (*ch <= 0x94) *ch -= 0x80;
+				else if (*ch <= 0x95) ;
+				// 96 -> 15 û:u circumflex, 97 -> 16 ù:u grave.
+				else if (*ch <= 0x97) *ch -= 0x81;
+				else if (*ch <= 0x98) ;
+				// 99 -> 17 Ö:O diaeresis,  9a -> 18 Ü:U diaeresis.
+				else if (*ch <= 0x9a) *ch -= 0x82;
+				else if (*ch <= 0x9f) ;
+				// a0 -> 19 á:a acute.
+				else if (*ch <= 0xa0) *ch -= 0x87;
+				// a1 -> 1d í:i acute,      a2 -> 1e ó:o acute,      a3 -> 1f ú:u acute.
+				else if (*ch <= 0xa3) *ch -= 0x84;
+				else if (*ch <= 0xe0) ;
+				// e1 -> 1c ß:s sharp.
+				else if (*ch <= 0xe1) *ch -= 0xc5;
+			}
+		}
```

Comment:
This whole block of code appears twice; it probably should be extracted to a function.